build/cmake/DefineOptions.cmake

```diff
@@ -62,8 +62,10 @@ endif()
 find_package(OpenSSL QUIET)
 CMAKE_DEPENDENT_OPTION(WITH_OPENSSL "Build with OpenSSL support" ON
                        "OPENSSL_FOUND" OFF)
-option(WITH_BOOSTTHREADS "Build with Boost thread support" OFF)
 option(WITH_STDTHREADS "Build with C++ std::thread support" OFF)
+CMAKE_DEPENDENT_OPTION(WITH_BOOSTTHREADS "Build with Boost threads support" ON
```

Comment:
According to http://www.cmake.org/cmake/help/v3.0/module/CMakeDependentOption.html

> This macro presents an option to the user only if a set of other conditions are true. 

This would hide the option itself unless the user set it in the first place, like a chicken out of no egg for GUI users.

How about literally changing the default value only:

    CMAKE_DEPENDENT_OPTION(WITH_BOOSTTHREADS "Build with Boost threads support" OFF
         "NOT WITH_STDTHREADS;Boost_FOUND" OFF)


About the config.h, I think your concern is right.
Long story short, it originates from my past patch, and in its current state I don't see any reason not to replace *add_definitions* with *set*.